docs/specs/consensus/README.md

```diff
@@ -0,0 +1,1111 @@
+
+The relative minimum window density of blocks `B1` and `B2` is defined as.
+
+```rust
+fn relativeMinWindowDensity(B1, B2) -> u32
+{
+    let max_slot = max(globalSlot(B1), globalSlot(B2))
+
+    // Grace-period rule
+    if max_slot < grace_period_end {
+        return B1.min_window_density
+    }
+
+    // Compute B1's window projected to max_slot
+    let projected_window = {
+        // Compute shift count
+        let shift_count = min(max(max_slot - B1.curr_global_slot - 1, 0), sub_windows_per_window)
+
+        // Initialize projected window
+        let projected_window = B1.sub_window_densities
+
+        // Ring-shift
+        let i = relativeSubWindow(B1.curr_global_slot)
+        while shift_count > 0 {
+          i = i + 1 mod sub_windows_per_window
+          projected_window[i] = 0
+          shift_count--;
+        }
+
+        return projected_window
+    }
+
+    // Compute projected window density
+    let projected_window_density = density(projected_window)
+
+    // Compute minimum window density
+    return min(B1.min_window_density, projected_window_density)
+}
+```
+
+This description was adopted to aid understanding, providing explanations where possible of why strategies have been adopted and what conditions are important.  In a production implementation consideration must be given to performance implications and another implementation may be desirable.
+
+# 6 Protocol
+
+This section specifies the consensus protocol in terms of events and how they MUST be implemented by a compatible peer.  The required events are:
+
+* [`Initialize consensus`](#61-initialize-consensus)
+* [`Select chain`](#62-select-chain)
+
+Additionally there are certain local data members that all peers MUST maintain in order to participate in consensus.
+
+| Parameter   | Description |
+| - | - |
+| `genesis_block` | The initial block in the blockchain |
+| `neighbors`     | Set of connections to neighboring peers |
+| `chains`        | Set of known (succinct) candidate chains |
+| `tip`           | Currently selected chain according to the chain selection algorithm (i.e. secure chain) |
+
+How these are represented is up to the implementation, but careful consideration must be given to scalability.
+
+In the following description we use _dot notation_ to refer the local data members of peers. For example, given peer `P`, we use `P.genesis_block` and `P.tip`, to refer to the genesis block and currently selected chain, respectively.
+
+## 6.1 Initialize consensus
+
+Things a peer MUST do to initialize consensus includes
+
+* Load the genesis block
+* Get head of the current chain
+* Bootstrap
+* Catchup
```

Comment:
Catchup is only necessary if the peer has disconnected candidate blocks after bootstrap.

Reply:
After bootstrap, you may have blocks you heard about over gossip, but do not have the dependencies for to link it to the point of finality (root of the frontier). Catchup is used to download those missing blocks. But my point here is that catchup, for us, is sort of an implementation detail related to how we designed out bootstrap (one could design a bootstrap which encapsulates this task, and the jobs we are doing in our catchup subsystem are actually part of the "bootstrapping" process, just not in our bootstrapping subsystem).